ocaml/xapi/xapi_pool.ml

```diff
@@ -615,6 +615,40 @@ let protect_exn f x =
 		Debug.log_backtrace e (Backtrace.get e);
 		None
 
+let create_or_get_pvs_server_on_master __context rpc session_id pvs_site_ref (pvs_server_ref, pvs_server) : API.ref_PVS_server =
+	let my_uuid = pvs_server.API.pVS_server_uuid in
+	let new_pvs_server_ref =
+		try Client.PVS_server.get_by_uuid ~rpc ~session_id ~uuid:my_uuid
+		with _ ->
+			debug "Found no PVS server with uuid = '%s' on the master, so creating one." my_uuid;
+			Client.PVS_server.introduce ~rpc ~session_id
+			~addresses:pvs_server.API.pVS_server_addresses
+			~first_port:pvs_server.API.pVS_server_first_port
+			~last_port:pvs_server.API.pVS_server_last_port
+			~site:pvs_site_ref
+	in
+	new_pvs_server_ref
+
+let create_or_get_pvs_site_on_master __context rpc session_id (pvs_site_ref, pvs_site) : API.ref_PVS_site =
+	let my_site_name = pvs_site.API.pVS_site_name in
+	let new_pvs_site_ref =
+		let expr = "field \"name\"=\"" ^ my_site_name ^ "\"" in
+		match Client.PVS_site.get_all_records_where ~rpc ~session_id ~expr with
+		| [] -> debug "Found no PVS site with name = '%s' on the master, so creating one." my_site_name;
+			let new_pvs_site = Client.PVS_site.introduce ~rpc ~session_id ~name:my_site_name in
+			(* Update PVS servers *)
+			let my_pvs_servers = Db.PVS_server.get_records_where ~__context ~expr:(
+				Eq(Field "site", Literal (Ref.string_of pvs_site_ref))) in
+			let (_ : API.ref_PVS_server option list) =
+				List.map (protect_exn (create_or_get_pvs_server_on_master __context rpc session_id new_pvs_site)) my_pvs_servers in
```

Comment:
This should not be a "create_and_get" function, because you always need to just "create". which I would just do inline.